### PR TITLE
Property delegates as custom attributes

### PR DESCRIPTION
**Pull request by DougGregor**
This is an implementation of [property delegates as custom attributes](https://github.com/DougGregor/swift-evolution/blob/property-delegates/proposals/NNNN-property-delegates.md), under discussion in this [pitch thread](https://forums.swift.org/t/pitch-2-property-delegates-by-custom-attributes/22855). The implementation is mostly complete, although there are some annoying known bugs:

* Property delegates cannot be used on local variables; only properties of types work well (`lazy` is similarly broken and always has been!)

### CODE REVIEW

**Comment by hartbit** [comment]
@DougGregor Would it be possible to have a toolchain created to play around with Property Delegates? I'd like to use them to improve a library Im working on as a way to test them out.

**Comment by DougGregor** [comment]
@hartbit let me make sure things are back in a working state now, and once they are I'll build a toolchain. I just got to the point where they work end-to-end, so long as you don't touch `@autoclosure`.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by hartbit** [comment]
That's a lot of toolchains 😂

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please please please please please build toolchain

**Comment by shahmishal** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by shahmishal** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci build toolchain

**Comment by shahmishal** [comment]
@swift-ci build toolchain Linux

**Comment by shahmishal** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please smoke test macOS

**Comment by hartbit** [comment]
With the implementation in the toolchain, the `DelayedMutable` example from the proposal fails to compile:

```swift
@propertyDelegate
struct DelayedMutable<Value> {
    private var _value: Value? = nil

    var value: Value {
        get {
            guard let value = _value else {
                fatalError("property accessed before being initialized")
            }
            return value
        }
        set {
            _value = newValue
        }
    }

    /// "Reset" the delegate so it can be initialized again.
    mutating func reset() {
        _value = nil
    }
}

class Foo {
    @DelayedMutable var x: Int

    init() {
        // We don't know "x" yet, and we don't have to set it
    } // error: Return from initializer without initializing all stored properties

    func initializeX(x: Int) {
        self.x = x // Will crash if 'self.x' is already initialized
    }

    func getX() -> Int {
        return x // Will crash if 'self.x' wasn't initialized
    }
}
```

Is this known?

**Comment by DougGregor** [comment]
@hartbit ah, thank you! that should be `@DelayedMutable() var x: Int`. I'll fix the proposal

**Comment by hartbit** [comment]
@DougGregor The `()` part of the syntax is surprising because the proposal does not talk about it much. When is it required?

**Comment by DougGregor** [comment]
@hartbit The `()` is there to call the no-parameter initializer. Without it, the delegate instance hasn't been initialized. There's no special case here

**Comment by DougGregor** [comment]
I got `Lazy` working, so time to build a new toolchain

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please build macOS toolchain

**Comment by DougGregor** [comment]
@swift-ci please build toolchain macOS

**Comment by hartbit** [comment]
@hartbit Can't the parentheses-less `@DelayedMutable` syntax be equivalent to calling the parameter-less initializer? That's the direction the Custom Attributes proposal is going towards.

**Comment by shahmishal** [comment]
@swift-ci please build toolchain macOS

**Comment by shahmishal** [comment]
@swift-ci please build toolchain macOS

**Comment by DougGregor** [comment]
@swift-ci build toolchain

**Comment by DougGregor** [comment]
@swift-ci please build toolchain macOS

**Comment by DougGregor** [comment]
> @hartbit Can't the parentheses-less `@DelayedMutable` syntax be equivalent to calling the parameter-less initializer? That's the direction the Custom Attributes proposal is going towards.

I don't think that's a good idea, because you would lose the ability to distinguish between "not initialized" and "initialized via the zero-parameter initializer".

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
Linux failure looks unrelated:

```
  /tmp/CFamilyTargets_ModuleMapGenerationCases.fQZYzj/CFamilyTargets_ModuleMapGenerationCases/Sources/Baz/main.swift:4:8: error: no such module 'UmbrellaHeaderFlat'
01:22:33     import UmbrellaHeaderFlat
01:22:33            ^
```

**Comment by tanner0101** [comment]
Is this error expected with the toolchain linked in [SE-0258](https://github.com/apple/swift-evolution/blob/master/proposals/0258-property-delegates.md)?

```
Ld /Users/tanner0101/dev/vapor/sandbox/pdel-spm/DerivedData/PropertyDelegates/Build/Products/Debug/PropertyDelegates.framework/Versions/A/PropertyDelegates normal x86_64 (in target: PropertyDelegates)
    cd /Users/tanner0101/dev/vapor/sandbox/pdel-spm
    export MACOSX_DEPLOYMENT_TARGET=10.10
    /Library/Developer/Toolchains/swift-PR-23701-287.xctoolchain/usr/bin/clang -arch x86_64 -dynamiclib -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk -L/Users/tanner0101/dev/vapor/sandbox/pdel-spm/DerivedData/PropertyDelegates/Build/Products/Debug -F/Users/tanner0101/dev/vapor/sandbox/pdel-spm/DerivedData/PropertyDelegates/Build/Products/Debug -F/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks -filelist /Users/tanner0101/dev/vapor/sandbox/pdel-spm/DerivedData/PropertyDelegates/Build/Intermediates.noindex/PropertyDelegates.build/Debug/PropertyDelegates.build/Objects-normal/x86_64/PropertyDelegates.LinkFileList -install_name @rpath/PropertyDelegates.framework/Versions/A/PropertyDelegates -Xlinker -rpath -Xlinker /usr/lib/swift -Xlinker -rpath -Xlinker /Library/Developer/Toolchains/swift-PR-23701-287.xctoolchain/usr/lib/swift/macosx -mmacosx-version-min=10.10 -Xlinker -object_path_lto -Xlinker /Users/tanner0101/dev/vapor/sandbox/pdel-spm/DerivedData/PropertyDelegates/Build/Intermediates.noindex/PropertyDelegates.build/Debug/PropertyDelegates.build/Objects-normal/x86_64/PropertyDelegates_lto.o -Xlinker -export_dynamic -Xlinker -no_deduplicate -fobjc-link-runtime -L/Library/Developer/Toolchains/swift-PR-23701-287.xctoolchain/usr/lib/swift/macosx -Xlinker -add_ast_path -Xlinker /Users/tanner0101/dev/vapor/sandbox/pdel-spm/DerivedData/PropertyDelegates/Build/Intermediates.noindex/PropertyDelegates.build/Debug/PropertyDelegates.build/Objects-normal/x86_64/PropertyDelegates.swiftmodule -Xlinker -dependency_info -Xlinker /Users/tanner0101/dev/vapor/sandbox/pdel-spm/DerivedData/PropertyDelegates/Build/Intermediates.noindex/PropertyDelegates.build/Debug/PropertyDelegates.build/Objects-normal/x86_64/PropertyDelegates_dependency_info.dat -o /Users/tanner0101/dev/vapor/sandbox/pdel-spm/DerivedData/PropertyDelegates/Build/Products/Debug/PropertyDelegates.framework/Versions/A/PropertyDelegates

ld: file not found: /Library/Developer/Toolchains/swift-PR-23701-287.xctoolchain/usr/lib/arc/libarclite_macosx.a
clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by hartbit** [comment]
@tanner0101 It's a known bug (https://bugs.swift.org/browse/SR-9972). As a workaround, you can symlink `libarclite_macosx.a` in the default toolchain to the PR toolchain:

```
$ ln -s /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/arc/libarclite_macosx.a /Library/Developer/Toolchains/swift-PR-23701-287.xctoolchain/usr/lib/arc/libarclite_macosx.a
```

PS: there's also a more recent toolchain for this PR, number 290

File: lib/Sema/TypeCheckPropertyDelegate.cpp [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+  PropertyDelegateTypeInfo result;
+  result.valueVar = valueVar;
+  result.initialValueInit = findInitialValueInit(ctx, nominal, valueVar);
+
+  return result;
+}
+
+llvm::Expected<CustomAttr *>
+AttachedPropertyDelegateRequest::evaluate(Evaluator &evaluator,
+                                          VarDecl *var) const {
```
Comment (rjmccall) [inline lib/Sema/TypeCheckPropertyDelegate.cpp]: This belongs in the next commit.

File: lib/Index/Index.cpp [hunk 2]
```diff
@@ -334,34 +334,47 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
...
+      Labels = MemberwiseInit->getFullName().getArgumentNames();
     } else if (auto *CallParent = dyn_cast_or_null<CallExpr>(getParentExpr())) {
       LabelLocs = CallParent->getArgumentLabelLocs();
+      Labels = CallParent->getArgumentLabels();
     }
 
     if (LabelLocs.empty())
       return;
 
+    assert(Labels.size() == LabelLocs.size());
```
Comment (benlangmuir) [inline lib/Index/Index.cpp]: This is "obviously correct" for the call case, but given that the compound name case is checking isValid() and skipping, I'm a bit suspicious.  Could you explain how we know this is true?

File: lib/Index/Index.cpp [hunk 3]
```diff
@@ -334,34 +334,47 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
...
         MemberwiseInit->getDeclContext()->getSelfNominalTypeDecl();
     if (!TypeContext || !shouldIndex(TypeContext, false))
       return;
 
-    auto LabelIt = LabelLocs.begin();
+    unsigned CurLabel = 0;
     for (auto Prop : TypeContext->getStoredProperties()) {
-      if (Prop->getParentInitializer() && Prop->isLet())
+      if (auto Original = Prop->getOriginalDelegatedProperty())
+        Prop = Original;
```
Comment (benlangmuir) [inline lib/Index/Index.cpp]: Could we add a test case for a delegated property?

File: lib/Sema/TypeCheckPropertyDelegate.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/Sema/TypeCheckPropertyDelegate.cpp]: Whoops, thank you!

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
And... another rebase is incoming.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by compnerd** [comment]
@DougGregor - *sigh* this broke the windows build :-( https://dev.azure.com/compnerd/windows-swift/_build/results?buildId=1697

**Comment by natecook1000** [comment]
@swift-ci Please built macOS toolchain

**Pull request merged**